File: ZDOpenSourceOCDemo/ZDOpenSourceOCDemo/Controllers/ViewController.m
//
//  ViewController.m
//  ZDOpenSourceOCDemo
//
//  Created by Zero.D.Saber on 2017/8/4.
//  Copyright © 2017年 Zero.D.Saber. All rights reserved.
//
//  3D Touch实践讲解 --解决开发中的困惑点: https://www.tuicool.com/articles/VjEZniY
//  3D Touch功能实现: https://www.tuicool.com/articles/qQZjuqF
//  3Dtouch 的实际应用详解(tableView中的应用): https://www.tuicool.com/articles/baEnqef

#import "ViewController.h"
#import <objc/runtime.h>
#import <objc/message.h>
#import <ReactiveObjC/ReactiveObjC.h>
#if __has_include(<KZPlayground/KZPPlayground.h>)
#import <KZPlayground/KZPPlayground.h>
#endif
#import <ZDToolKit/ZDToolKit.h>
#import <JLRoutes/JLRoutes.h>
#import <ZIKRouter/ZIKViewRouter.h>

@interface ViewController () <UITableViewDataSource, UITableViewDelegate
#if __IPHONE_OS_VERSION_MAX_ALLOWED >= __IPHONE_9_0
, UIViewControllerPreviewingDelegate
#endif
>
@property (nonatomic, weak  ) IBOutlet UIBarButtonItem *rightItem;
@property (nonatomic, weak  ) IBOutlet UITableView *tableView;
@property (nonatomic, strong) NSArray<NSString *> *dataSource;
@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    [self setup];
}

- (void)setup {
    [self setupData];
    [self setupUI];
}

- (void)setupData {
    NSArray<NSString *> *data = @[
      @"YogaKit",
      @"YogaKitList",
      @"IGListKit",
      @"Texture",
      @"JLRoute",
      @"Lottie",
    ];
    self.dataSource = data;
}

- (void)setupUI {
    self.view.backgroundColor = [UIColor whiteColor];
    
    [self.tableView registerClass:[UITableViewCell class] forCellReuseIdentifier:NSStringFromClass([UITableViewCell class])];
}

#pragma mark - UITableViewDatasource && UITableViewDelegate

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.dataSource.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:NSStringFromClass([UITableViewCell class]) forIndexPath:indexPath];
    cell.selectionStyle = UITableViewCellSelectionStyleNone;
    cell.textLabel.text = self.dataSource[indexPath.row];
    if (self.traitCollection.forceTouchCapability == UIForceTouchCapabilityAvailable) {
        [self registerForPreviewingWithDelegate:self sourceView:cell];
    }
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    NSString *text = self.dataSource[indexPath.row];
    
    BOOL useZIKRouter = YES;
    if (useZIKRouter) {
        NSString *routerName = [text stringByAppendingString:@"ViewRouter"];
        Class aClass = objc_getClass(routerName.UTF8String);
        [aClass performPath:ZIKViewRoutePath.pushFrom(self)];
    }
    else {
        NSString *vcName = [text stringByAppendingString:@"ViewController"];
        NSString *url = [NSString stringWithFormat:@"ZDRoute://push/%@?arg1=hello&arg2=world", vcName];// @"ZDRoute://push?viewController=%@arg1=hello&arg2=world"
#if 1
        [[JLRoutes routesForScheme:@"ZDRoute"] routeURL:[NSURL URLWithString:url]];
#else
        [[UIApplication sharedApplication] openURL:[NSURL URLWithString:url] options:@{UIApplicationOpenURLOptionUniversalLinksOnly : @NO} completionHandler:^(BOOL success) {
        //
        }];
#endif
    }
}

#if __IPHONE_OS_VERSION_MAX_ALLOWED >= __IPHONE_9_0
#pragma mark - UIViewControllerPreviewingDelegate
// peek
- (UIViewController *)previewingContext:(id<UIViewControllerPreviewing>)previewingContext viewControllerForLocation:(CGPoint)location {
    NSString *text = ((UITableViewCell *)previewingContext.sourceView).textLabel.text;
    Class aClass = objc_getClass([text stringByAppendingString:@"ViewController"].UTF8String);
    return [aClass new];
}

// pop
- (void)previewingContext:(id<UIViewControllerPreviewing>)previewingContext commitViewController:(UIViewController *)viewControllerToCommit {
    viewControllerToCommit.hidesBottomBarWhenPushed = YES;
    [self showViewController:viewControllerToCommit sender:self];
}
#endif

#pragma mark -

- (IBAction)present:(UIBarButtonItem *)sender {
#if __has_include(<KZPlayground/KZPPlayground.h>)
    [self.navigationController showViewController:[KZPPlaygroundViewController playgroundViewController] sender:sender];
#endif
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

@end

File: ZDOpenSourceSwiftDemo/ZDOpenSourceSwiftDemo/ViewControllers/ViewController.swift
//
//  ViewController.swift
//  ZDOpenSourceSwiftDemo
//
//  Created by Zero.D.Saber on 2017/8/4.
//  Copyright © 2017年 Zero.D.Saber. All rights reserved.
//

import UIKit
import ObjectiveC.runtime

class ViewController: ZDBaseViewController {
    var dataSource: [AnyClass] = []
    
    let ZDReuseCellIndentifier = "ReuseCell"
    // 每当这个类的新实例被创建时,这个闭包就会被调用,而闭包的返回值就会当做默认值赋值给这个属性
    // 添加lazy属性的话就相当于懒加载了,而且必须声明属性是变量;另外我们需要显示的指定属性的类型,并使用一个可以对这个属性进行赋值的语句来再首次访问时运行;
    // 因为实例在没有值的情况下以某种方式被初始化,然后在被访问时改变自己的值,所以要求该属性是可变的;
    // let 被声明在全局作用域下 或者 被声明为一个类型属性(声明为static let,而非声明为实例属性)的常量是自动具有惰性的(还是线程安全的)
    lazy private var tableView: UITableView = {
        let view = UITableView(frame: .zero, style: UITableView.Style.plain)
        view.register(UITableViewCell.self, forCellReuseIdentifier: "ReuseCell")
        view.separatorInset = .zero
        return view
    }()
    
    // 以下写法 <=> Objective-C中的GNU-C写法
    let button1: UIButton = {
        let button = UIButton(type: UIButton.ButtonType.custom)
        button.addTarget(nil, action: #selector(ViewController.click), for: .touchUpInside)
        return button
    }()
    
    lazy var button2: UIButton = {
        let button = UIButton(type: UIButton.ButtonType.custom)
        button.addTarget(self, action: #selector(ViewController.click), for: .touchUpInside)
        return button
    }()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        setup()
    }

    private func setup() {
        setupUI()
        setupData()
    }
    
    private func setupUI() {
        view.backgroundColor = .cyan
        self.edgesForExtendedLayout = UIRectEdge(rawValue: 0)
        
        tableView.dataSource = self
        tableView.delegate = self
        view.addSubview(tableView)
        
        // add constraint
        tableView.translatesAutoresizingMaskIntoConstraints = false
        let viewDict = ["tableView": tableView]
        view.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "H:|-0-[tableView]-0-|", options: .directionLeadingToTrailing, metrics: nil, views: viewDict))
        view.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "V:|-0-[tableView]-0-|", options: .directionLeadingToTrailing, metrics: nil, views: viewDict))
    }
    
    private func setupData() {
        let aClasses: [UIViewController.Type] = [YogaKitViewController.self, ZDFeedController.self, ZDMovieListController.self]
        
        for _ in 0...50 {
            let num: UInt32 = arc4random_uniform(UInt32(aClasses.count))
            dataSource.append(aClasses[Int(num)])
        }
        tableView.reloadData()
    }
    
    @objc func click() {}
    
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
}

extension ViewController: UITableViewDataSource {
    func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return dataSource.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: ZDReuseCellIndentifier)
        cell?.selectionStyle = .none
        cell?.layoutMargins = .zero
        return cell ?? UITableViewCell(style: .default, reuseIdentifier: NSStringFromClass(UITableViewCell.self))
    }
}

extension ViewController: UITableViewDelegate {
    func tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell, forRowAt indexPath: IndexPath) {
        cell.textLabel?.text = "\(self.dataSource[indexPath.row])"
        cell.detailTextLabel?.text = "第\(indexPath.row + 1)行"
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        defer { // 出了当前作用域后执行,相当于RAC中的onExit宏
            tableView.deselectRow(at: indexPath, animated: true)
        }

        let aClass: UIViewController.Type = dataSource[indexPath.row] as! UIViewController.Type
        print("\(aClass.self)")
        navigationController?.show(aClass.init(), sender: nil)
    }
}

















